### ISSUE TITLE
MosesTokenizer is skipping the final fullstop.

### ISSUE THREAD

**Issue by alvations** [nltk/nltk#1616]
MosesTokenizer is skipping the final fullstop.

```python
>>> from nltk.tokenize.moses import MosesTokenizer
>>> m = MosesTokenizer()
>>> m.tokenize('abc def.')
[u'abc', u'def.']
>>> m.penn_tokenize('abc def.')
[u'abc', u'def', u'.']
```

But MosesTokenizer's penn_tokenize works because of https://github.com/nltk/nltk/blob/develop/nltk/tokenize/moses.py#L103

While the default tokenizer relies on the `handles_nonbreaking_prefixes()` to handle the final fullstop: https://github.com/nltk/nltk/blob/develop/nltk/tokenize/moses.py#L284

The conditions at https://github.com/nltk/nltk/blob/develop/nltk/tokenize/moses.py#L274 is wrong since it's not checking what Moses is actually doing on https://github.com/alvations/mosesdecoder/blob/master/scripts/tokenizer/tokenizer.perl#L333

The correct implementation should be:

```python
    def handles_nonbreaking_prefixes(self, text):
        # Splits the text into tokens to check for nonbreaking prefixes.
        tokens = text.split()
        num_tokens = len(tokens)
        for i, token in enumerate(tokens):
            # Checks if token ends with a fullstop.
            token_ends_with_period = re.findall(r'^(\S+)\.$', token)
            if token_ends_with_period:
                prefix = token_ends_with_period.group(1)
                # Checks for 3 conditions if
                # i.   the prefix contains a fullstop and
                #      any char in the prefix is within the IsAlpha charset
                # ii.  the prefix is in the list of nonbreaking prefixes and
                #      does not contain #NUMERIC_ONLY#
                # iii. the token is not the last token and that the
                #      next token contains all lowercase.
                if ( ('.' in prefix and self.isalpha(prefix)) or
                     (prefix in self.NONBREAKING_PREFIXES and
                      prefix not in self.NUMERIC_ONLY_PREFIXES) or
                     (i != num_tokens-1 and self.islower(tokens[i+1])) ):
                    pass # No change to the token.
                # Checks if the prefix is in NUMERIC_ONLY_PREFIXES
                # and ensures that the next word is a digit.
                elif (prefix in self.NUMERIC_ONLY_PREFIXES and
                      re.search(r'^[0-9]+', tokens[i+1])):
                    pass # No change to the token.
                else: # Otherwise, adds a space after the tokens before a dot.
                    tokens[i] = prefix + ' .'
        return " ".join(tokens) # Stitch the tokens back.
```

After the patch:

```
$ python -c "from nltk.tokenize.moses import MosesTokenizer; m = MosesTokenizer(); print(m.tokenize('abc def.'))"
[u'abc', u'def', u'.']
```

**Comment by alvations** [comment]
Patched on #1702

**Comment by Henry-E** [comment]
@alvations This breaks when the last token in the list has a full stop.

```  
File "/home/henrye/anaconda3/envs/pytorch36/lib/python3.6/sitepackages/nltk/tokenize/moses.py", 
line 289, in handles_nonbreaking_prefixes
    re.search(r'^[0-9]+', tokens[i+1])):
IndexError: list index out of range```

**Comment by Henry-E** [comment]
Hmmm, it must be the version of NLTK that I have but this appears to be fixed in the current repo 
https://github.com/nltk/nltk/commit/22cbc3b9c1ca0bbe41afaf80bcfaa5e178476df4

**Comment by alvations** [comment]
@Henry-E Thank you for reporting the issue. Yes, this has been fixed in 22cbc3b. The patch should be available in the next minor version release of NLTK. 

Meanwhile, feel free to install from the master branch if you would like the fix urgently but do note that it might be subjected errors/breakage, e.g. 

```
pip install -U https://github.com/nltk/nltk/archive/develop.zip
```

**Comment by Henry-E** [comment]
Thanks for letting me know. For a small bug like this I just manually added the line to the module and now it's working again properly.
